File: backend/scripts/test-ingestion-loader.php
<?php

declare(strict_types=1);

/**
 * Harness test for AgentTeam\Services\IngestionLoader — the interpreter's
 * file-read step (recursive list_files -> read_file -> decode -> text).
 *
 * Plain-assert style (same as test-vector-mcp-store.php). The MCP transport
 * (list_files / read_file) is INJECTED as callables, so the recursive
 * enumeration and decoding are tested with no live MCP server. Run with:
 *
 *     php backend/scripts/test-ingestion-loader.php
 */

require __DIR__ . '/../vendor/autoload.php';

use AgentTeam\Services\IngestionLoader;

$FIXTURES = __DIR__ . '/../tests/fixtures';
$failures = 0;
$checks = 0;

function check(string $label, bool $cond): void
{
    global $failures, $checks;
    $checks++;
    if ($cond) {
        echo "  PASS  $label\n";
    } else {
        $failures++;
        echo "  FAIL  $label\n";
    }
}

// ---------------------------------------------------------------------------
// Fixtures: a fake list_files backed by an in-memory tree, a fake read_file
// backed by a blob map, and a minimal real .docx (zip) built in-process so
// decode is tested against the real libraries, not a mock.
// ---------------------------------------------------------------------------
function fakeListFiles(array $tree): callable
{
    return fn(string $provider, string $path): array => ['files' => $tree[$path] ?? []];
}

function fakeReadFile(array $blobs): callable
{
    return fn(string $provider, string $fileId): string => $blobs[$fileId];
}

function f(string $path): array
{
    return ['id' => $path, 'name' => basename($path), 'type' => 'file'];
}

function d(string $path): array
{
    return ['id' => $path, 'name' => basename($path), 'type' => 'folder'];
}

function makeDocx(string $text): string
{
    $tmp = tempnam(sys_get_temp_dir(), 'docx');
    $z = new ZipArchive();
    $z->open($tmp, ZipArchive::OVERWRITE);
    $z->addFromString(
        '[Content_Types].xml',
        '<?xml version="1.0"?><Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">'
        . '<Override PartName="/word/document.xml" ContentType="application/vnd.openxmlformats-officedocument.'
        . 'wordprocessingml.document.main+xml"/></Types>'
    );
    $z->addFromString(
        'word/document.xml',
        '<?xml version="1.0"?><w:document xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main">'
        . "<w:body><w:p><w:r><w:t>$text</w:t></w:r></w:p></w:body></w:document>"
    );
    $z->close();
    $bytes = file_get_contents($tmp);
    unlink($tmp);
    return $bytes;
}

// ===========================================================================
// detectType
// ===========================================================================
check('detectType pdf', IngestionLoader::detectType('report.pdf') === 'pdf');
check('detectType PDF (case)', IngestionLoader::detectType('notes.PDF') === 'pdf');
check('detectType docx', IngestionLoader::detectType('memo.docx') === 'word');
check('detectType doc', IngestionLoader::detectType('memo.doc') === 'word');
check('detectType txt', IngestionLoader::detectType('readme.txt') === 'text');
check('detectType csv', IngestionLoader::detectType('data.csv') === 'csv');
check('detectType path', IngestionLoader::detectType('a/b/deep.pdf') === 'pdf');
check('detectType html', IngestionLoader::detectType('page.html') === 'html');
check('detectType htm', IngestionLoader::detectType('page.htm') === 'html');
check('detectType unsupported', IngestionLoader::detectType('archive.zip') === null);
check('detectType no ext', IngestionLoader::detectType('noext') === null);

// ===========================================================================
// decode
// ===========================================================================
check('decode text', IngestionLoader::decode('readme.txt', 'hello world') === 'hello world');
check('decode csv', IngestionLoader::decode('data.csv', "x,y\n1,2\n3,4") === "x,y\n1,2\n3,4");
check('decode docx', str_contains(IngestionLoader::decode('memo.docx', makeDocx('Hello DOCX world')), 'Hello DOCX world'));
check('decode html', str_contains(IngestionLoader::decode('a.html', '<html><body><h1>Hello HTML world</h1><p>Body text.</p></body></html>'), 'Hello HTML world'));
check('decode html strips script', !str_contains(IngestionLoader::decode('a.html', '<html><body><script>var secret=1;</script><p>Visible only.</p></body></html>'), 'secret'));
check('decode pdf', str_contains(IngestionLoader::decode('report.pdf', file_get_contents("$FIXTURES/loader-sample.pdf")), 'Hello PDF world'));
check('decode auto routes pdf', str_contains(IngestionLoader::decode('report.pdf', file_get_contents("$FIXTURES/loader-sample.pdf"), 'auto'), 'Hello PDF world'));
check('decode auto routes text', IngestionLoader::decode('notes.txt', 'plain', 'auto') === 'plain');

$threw = false;
try {
    IngestionLoader::decode('archive.zip', 'PK');
} catch (\InvalidArgumentException $e) {
    $threw = true;
}
check('decode unsupported throws', $threw);

// ===========================================================================
// enumerateFiles
// ===========================================================================
// $allowedTypes is the checkbox filter: [] = no restriction (all supported).
$single = IngestionLoader::enumerateFiles(fakeListFiles([]), 'local', 'doc.pdf', [], false);
check('enumerate single count', count($single) === 1);
check('enumerate single file_id', $single[0]['file_id'] === 'doc.pdf');
check('enumerate single source', $single[0]['source'] === 'doc.pdf');
check('enumerate single doc_type', $single[0]['doc_type'] === 'pdf');
check('enumerate single provider', $single[0]['provider'] === 'local');

// Single file whose type is NOT in the checked set -> filtered out.
$excluded = IngestionLoader::enumerateFiles(fakeListFiles([]), 'local', 'doc.pdf', ['word'], false);
check('enumerate single excluded by filter', $excluded === []);
// ...and kept when its type IS checked.
$included = IngestionLoader::enumerateFiles(fakeListFiles([]), 'local', 'doc.pdf', ['pdf', 'word'], false);
check('enumerate single kept when type checked', count($included) === 1 && $included[0]['doc_type'] === 'pdf');

$tree = [
    'root'           => [f('root/a.pdf'), f('root/notes.txt'), f('root/image.png'), d('root/sub')],
    'root/sub'       => [f('root/sub/b.docx'), d('root/sub/deep')],
    'root/sub/deep'  => [f('root/sub/deep/c.csv')],
];
$all = IngestionLoader::enumerateFiles(fakeListFiles($tree), 'local', 'root', [], true);
$bySource = [];
foreach ($all as $it) {
    $bySource[$it['source']] = $it['doc_type'];
}
check('enumerate folder recursive, no filter = all supported', $bySource === [
    'root/a.pdf' => 'pdf',
    'root/notes.txt' => 'text',
    'root/sub/b.docx' => 'word',
    'root/sub/deep/c.csv' => 'csv',
]);
check('enumerate folder skips unsupported', !isset($bySource['root/image.png']));

$onlyPdf = IngestionLoader::enumerateFiles(fakeListFiles($tree), 'local', 'root', ['pdf'], true);
check('enumerate folder filter single type', array_map(fn($i) => $i['source'], $onlyPdf) === ['root/a.pdf']);

// The user's example: check Word + PDF -> only .docx and .pdf are processed.
$pdfWord = IngestionLoader::enumerateFiles(fakeListFiles($tree), 'local', 'root', ['pdf', 'word'], true);
$pwSources = array_map(fn($i) => $i['source'], $pdfWord);
sort($pwSources);
check('enumerate folder filter PDF+Word', $pwSources === ['root/a.pdf', 'root/sub/b.docx']);

$cycleTree = [
    'root'      => [f('root/a.pdf'), d('root/loop')],
    'root/loop' => [f('root/loop/b.txt'), d('root')],   // back-edge
];
$cyc = IngestionLoader::enumerateFiles(fakeListFiles($cycleTree), 'local', 'root', [], true);
$cycSources = array_map(fn($i) => $i['source'], $cyc);
sort($cycSources);
check('enumerate cycle guard', $cycSources === ['root/a.pdf', 'root/loop/b.txt']);

// ===========================================================================
// loadFile / loadDocuments
// ===========================================================================
$rf = fakeReadFile(['root/notes.txt' => 'hello from disk']);
$desc = IngestionLoader::enumerateFiles(fakeListFiles([]), 'local', 'root/notes.txt', [], false)[0];
check('loadFile reads and decodes', IngestionLoader::loadFile($rf, $desc) === 'hello from disk');

$docTree = [
    'root'     => [f('root/a.txt'), f('root/b.csv'), f('root/skip.png'), d('root/sub')],
    'root/sub' => [f('root/sub/c.txt')],
];
$blobs = [
    'root/a.txt'      => 'alpha',
    'root/b.csv'      => "x,y\n1,2",
    'root/sub/c.txt'  => 'gamma',
];
$docs = iterator_to_array(IngestionLoader::loadDocuments(
    fakeListFiles($docTree),
    fakeReadFile($blobs),
    'local',
    'root',
    [],
    true
), false);
$docMap = [];
foreach ($docs as $doc) {
    $docMap[$doc['source']] = $doc['text'];
}
check('loadDocuments yields text+source', $docMap === [
    'root/a.txt' => 'alpha',
    'root/b.csv' => "x,y\n1,2",
    'root/sub/c.txt' => 'gamma',
]);

$pdfBlobs = ['root/r.pdf' => file_get_contents("$FIXTURES/loader-sample.pdf")];
$pdfTree = ['root' => [f('root/r.pdf')]];
$pdfDocs = iterator_to_array(IngestionLoader::loadDocuments(
    fakeListFiles($pdfTree),
    fakeReadFile($pdfBlobs),
    'local',
    'root',
    [],
    true
), false);
check('loadDocuments pdf end-to-end',
    count($pdfDocs) === 1
    && str_contains($pdfDocs[0]['text'], 'Hello PDF world')
    && $pdfDocs[0]['source'] === 'root/r.pdf');

// ===========================================================================
// readRound — one file's text per round (what the Output tab shows)
// ===========================================================================
$rrTree = [
    'root'     => [f('root/a.txt'), f('root/b.csv'), f('root/skip.png'), d('root/sub')],
    'root/sub' => [f('root/sub/c.txt')],
];
$rrBlobs = [
    'root/a.txt'     => 'alpha',
    'root/b.csv'     => "x,y\n1,2",
    'root/sub/c.txt' => 'gamma',
];
$lf = fakeListFiles($rrTree);
$rf = fakeReadFile($rrBlobs);

$r0 = IngestionLoader::readRound($lf, $rf, 'local', 'root', [], true, 0);
check('readRound count = supported files only', $r0['count'] === 3);
check('readRound exposes ordered sources', $r0['sources'] === ['root/a.txt', 'root/b.csv', 'root/sub/c.txt']);
check('readRound round 0 = first file text', $r0['current']['source'] === 'root/a.txt' && $r0['current']['text'] === 'alpha' && $r0['current']['type'] === 'text');

$r2 = IngestionLoader::readRound($lf, $rf, 'local', 'root', [], true, 2);
check('readRound advances to round 2', $r2['cursor'] === 2 && $r2['current']['source'] === 'root/sub/c.txt' && $r2['current']['text'] === 'gamma');

$rEnd = IngestionLoader::readRound($lf, $rf, 'local', 'root', [], true, 3);
check('readRound past the end = exhausted (no current)', $rEnd['count'] === 3 && $rEnd['current'] === null);

// Long text is clipped for the round display.
$rClip = IngestionLoader::readRound(
    fakeListFiles(['r' => [f('r/big.txt')]]),
    fakeReadFile(['r/big.txt' => str_repeat('z', 100)]),
    'local', 'r', [], true, 0, 10
);
check('readRound clips long text', $rClip['current']['clipped'] === true && mb_strlen($rClip['current']['text']) === 10);

// A decode/read failure surfaces on the round, not as a fatal.
$boomRead = function (string $provider, string $fileId) {
    if ($fileId === 'r/bad.txt') {
        throw new \RuntimeException('read blew up');
    }
    return 'ok';
};
$rErr = IngestionLoader::readRound(
    fakeListFiles(['r' => [f('r/bad.txt'), f('r/good.txt')]]),
    $boomRead, 'local', 'r', [], true, 0
);
check('readRound surfaces per-file error', isset($rErr['current']['error']) && $rErr['current']['source'] === 'r/bad.txt');

// ---------------------------------------------------------------------------
echo "\n$checks checks, $failures failures\n";
exit($failures === 0 ? 0 : 1);

File: backend_python/tests/unit/test_ingestion_loader.py
"""Ported from backend/scripts/test-ingestion-loader.php -- WITH A DEVIATION:
running that script against the CURRENT IngestionLoader.php
(`php backend/scripts/test-ingestion-loader.php`, 2026-09-07) fatals at line
101 with `Call to undefined method IngestionLoader::decode()`. The PHP class
no longer has a decode() method: its docblock (IngestionLoader.php:139-146)
says the legacy UniversalFS raw-bytes decode path (pdf/docx/html extraction)
was REMOVED once every storage server became a langfs/LangChain loader that
extracts text server-side; `loadFile()` now requires the injected `readFile`
callable to already return {'is_text': True, 'data': str} and raises
otherwise. So per "PHP wins over the brief": this test file ports
detectType/enumerateFiles/loadFile/loadDocuments/readRound against the
CURRENT source (verified via `php -r` against the real class, not the stale
script) and does NOT port decode() -- there is nothing to port; it doesn't
exist. See task-5-report.md for the full deviation writeup.
"""
from __future__ import annotations

from app.agent_team.services.ingestion_loader import IngestionLoader


def _f(path: str) -> dict:
    return {'id': path, 'name': path.rsplit('/', 1)[-1], 'type': 'file'}


def _d(path: str) -> dict:
    return {'id': path, 'name': path.rsplit('/', 1)[-1], 'type': 'folder'}


def _fake_list_files(tree: dict):
    return lambda provider, path: {'files': tree.get(path, [])}


def _fake_read_file(blobs: dict):
    """Current contract: readFile returns {'is_text': True, 'data': str} on
    success (langfs already extracted the text); a missing/unreadable file
    returns {'is_text': False} (loadFile then raises)."""
    def rf(provider, fileId):
        if fileId not in blobs:
            return {'is_text': False}
        return {'is_text': True, 'data': blobs[fileId]}
    return rf


# ---------------------------------------------------------------------------
# detectType (pinned via `php -r`, 2026-09-07 -- unchanged from the stale
# script's assertions, lines 86-96, which DID run before the fatal)
# ---------------------------------------------------------------------------

def test_detect_type_pdf():
    assert IngestionLoader.detectType('report.pdf') == 'pdf'


def test_detect_type_pdf_case_insensitive():
    assert IngestionLoader.detectType('notes.PDF') == 'pdf'


def test_detect_type_docx():
    assert IngestionLoader.detectType('memo.docx') == 'word'


def test_detect_type_doc():
    assert IngestionLoader.detectType('memo.doc') == 'word'


def test_detect_type_txt():
    assert IngestionLoader.detectType('readme.txt') == 'text'


def test_detect_type_csv():
    assert IngestionLoader.detectType('data.csv') == 'csv'


def test_detect_type_with_path():
    assert IngestionLoader.detectType('a/b/deep.pdf') == 'pdf'


def test_detect_type_html():
    assert IngestionLoader.detectType('page.html') == 'html'


def test_detect_type_htm():
    assert IngestionLoader.detectType('page.htm') == 'html'


def test_detect_type_unsupported():
    assert IngestionLoader.detectType('archive.zip') is None


def test_detect_type_no_extension():
    assert IngestionLoader.detectType('noext') is None


# ---------------------------------------------------------------------------
# enumerateFiles (pinned via `php -r`, 2026-09-07 -- this section never ran
# in the stale script, so it is verified fresh here against current source)
# ---------------------------------------------------------------------------

def test_enumerate_single_file():
    single = IngestionLoader.enumerateFiles(_fake_list_files({}), 'local', 'doc.pdf', [], False)
    assert single == [{
        'provider': 'local', 'file_id': 'doc.pdf', 'name': 'doc.pdf',
        'source': 'doc.pdf', 'doc_type': 'pdf',
    }]


def test_enumerate_single_excluded_by_filter():
    excluded = IngestionLoader.enumerateFiles(_fake_list_files({}), 'local', 'doc.pdf', ['word'], False)
    assert excluded == []


def test_enumerate_single_kept_when_type_checked():
    included = IngestionLoader.enumerateFiles(_fake_list_files({}), 'local', 'doc.pdf', ['pdf', 'word'], False)
    assert len(included) == 1 and included[0]['doc_type'] == 'pdf'


def _tree():
    return {
        'root': [_f('root/a.pdf'), _f('root/notes.txt'), _f('root/image.png'), _d('root/sub')],
        'root/sub': [_f('root/sub/b.docx'), _d('root/sub/deep')],
        'root/sub/deep': [_f('root/sub/deep/c.csv')],
    }


def test_enumerate_folder_recursive_no_filter_keeps_all_supported():
    all_ = IngestionLoader.enumerateFiles(_fake_list_files(_tree()), 'local', 'root', [], True)
    by_source = {it['source']: it['doc_type'] for it in all_}
    assert by_source == {
        'root/a.pdf': 'pdf',
        'root/notes.txt': 'text',
        'root/sub/b.docx': 'word',
        'root/sub/deep/c.csv': 'csv',
    }


def test_enumerate_folder_skips_unsupported():
    all_ = IngestionLoader.enumerateFiles(_fake_list_files(_tree()), 'local', 'root', [], True)
    assert 'root/image.png' not in {it['source'] for it in all_}


def test_enumerate_folder_filter_single_type():
    only_pdf = IngestionLoader.enumerateFiles(_fake_list_files(_tree()), 'local', 'root', ['pdf'], True)
    assert [i['source'] for i in only_pdf] == ['root/a.pdf']


def test_enumerate_folder_filter_pdf_and_word():
    pdf_word = IngestionLoader.enumerateFiles(_fake_list_files(_tree()), 'local', 'root', ['pdf', 'word'], True)
    assert sorted(i['source'] for i in pdf_word) == ['root/a.pdf', 'root/sub/b.docx']


def test_enumerate_cycle_guard():
    cycle_tree = {
        'root': [_f('root/a.pdf'), _d('root/loop')],
        'root/loop': [_f('root/loop/b.txt'), _d('root')],  # back-edge
    }
    cyc = IngestionLoader.enumerateFiles(_fake_list_files(cycle_tree), 'local', 'root', [], True)
    assert sorted(i['source'] for i in cyc) == ['root/a.pdf', 'root/loop/b.txt']


def test_enumerate_max_depth_cap():
    deep_tree = {}
    for i in range(5):
        deep_tree[f'lvl{i}'] = [_d(f'lvl{i + 1}')]
    deep_tree['lvl5'] = [_f('lvl5/x.txt')]
    deep = IngestionLoader.enumerateFiles(_fake_list_files(deep_tree), 'local', 'lvl0', [], True, 3)
    assert deep == []  # lvl4 (depth 4) is never listed, so lvl5/x.txt is never reached


# ---------------------------------------------------------------------------
# loadFile (pinned via `php -r`, 2026-09-07 -- reflects the CURRENT contract:
# readFile must return {'is_text': True, 'data': str}, or loadFile raises)
# ---------------------------------------------------------------------------

def test_load_file_reads_and_returns_text():
    rf = _fake_read_file({'root/notes.txt': 'hello from disk'})
    desc = IngestionLoader.enumerateFiles(_fake_list_files({}), 'local', 'root/notes.txt', [], False)[0]
    assert IngestionLoader.loadFile(rf, desc) == 'hello from disk'


def test_load_file_raises_when_read_file_does_not_return_text():
    rf = _fake_read_file({})  # any fileId returns {'is_text': False}
    desc = {'provider': 'local', 'file_id': 'missing.txt', 'name': 'missing.txt', 'doc_type': 'text'}
    try:
        IngestionLoader.loadFile(rf, desc)
        assert False, 'expected RuntimeError'
    except RuntimeError as e:
        assert str(e) == (
            "read_file did not return extracted text for 'missing.txt' — "
            "the legacy raw-bytes decode path was removed; use a langfs storage server (format=text)."
        )


# ---------------------------------------------------------------------------
# loadDocuments (generator)
# ---------------------------------------------------------------------------

def test_load_documents_yields_text_and_source():
    doc_tree = {
        'root': [_f('root/a.txt'), _f('root/b.csv'), _f('root/skip.png'), _d('root/sub')],
        'root/sub': [_f('root/sub/c.txt')],
    }
    blobs = {'root/a.txt': 'alpha', 'root/b.csv': 'x,y\n1,2', 'root/sub/c.txt': 'gamma'}
    docs = list(IngestionLoader.loadDocuments(_fake_list_files(doc_tree), _fake_read_file(blobs), 'local', 'root', [], True))
    doc_map = {d['source']: d['text'] for d in docs}
    assert doc_map == {
        'root/a.txt': 'alpha',
        'root/b.csv': 'x,y\n1,2',
        'root/sub/c.txt': 'gamma',
    }


# ---------------------------------------------------------------------------
# readRound -- one file's text per round (the Output tab)
# ---------------------------------------------------------------------------

def _rr_setup():
    rr_tree = {
        'root': [_f('root/a.txt'), _f('root/b.csv'), _f('root/skip.png'), _d('root/sub')],
        'root/sub': [_f('root/sub/c.txt')],
    }
    rr_blobs = {'root/a.txt': 'alpha', 'root/b.csv': 'x,y\n1,2', 'root/sub/c.txt': 'gamma'}
    return _fake_list_files(rr_tree), _fake_read_file(rr_blobs)


def test_read_round_count_is_supported_files_only():
    lf, rf = _rr_setup()
    r0 = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, 0)
    assert r0['count'] == 3


def test_read_round_exposes_ordered_sources():
    lf, rf = _rr_setup()
    r0 = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, 0)
    assert r0['sources'] == ['root/a.txt', 'root/b.csv', 'root/sub/c.txt']


def test_read_round_zero_is_first_file_text():
    lf, rf = _rr_setup()
    r0 = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, 0)
    assert r0['current']['source'] == 'root/a.txt'
    assert r0['current']['text'] == 'alpha'
    assert r0['current']['type'] == 'text'
    assert r0['current']['chars'] == 5
    assert r0['current']['clipped'] is False


def test_read_round_advances():
    lf, rf = _rr_setup()
    r2 = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, 2)
    assert r2['cursor'] == 2
    assert r2['current']['source'] == 'root/sub/c.txt'
    assert r2['current']['text'] == 'gamma'


def test_read_round_past_end_is_exhausted():
    lf, rf = _rr_setup()
    r_end = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, 3)
    assert r_end['count'] == 3 and r_end['current'] is None


def test_read_round_negative_cursor_is_exhausted():
    lf, rf = _rr_setup()
    r_neg = IngestionLoader.readRound(lf, rf, 'local', 'root', [], True, -1)
    assert r_neg['count'] == 3 and r_neg['current'] is None


def test_read_round_no_files_matched():
    _, rf = _rr_setup()
    r_empty = IngestionLoader.readRound(_fake_list_files({}), rf, 'local', 'nope.txt', ['word'], False, 0)
    assert r_empty == {'count': 0, 'cursor': 0, 'sources': [], 'current': None}


def test_read_round_clips_long_text():
    r_clip = IngestionLoader.readRound(
        _fake_list_files({'r': [_f('r/big.txt')]}),
        _fake_read_file({'r/big.txt': 'z' * 100}),
        'local', 'r', [], True, 0, 10,
    )
    assert r_clip['current']['clipped'] is True
    assert len(r_clip['current']['text']) == 10
    assert r_clip['current']['text'] == 'z' * 10


def test_read_round_clips_by_code_points_not_bytes():
    # 30 code points, each 3 UTF-8 bytes -- clip must cut at 12 CHARACTERS.
    mb_text = '日本語' * 10
    r_clip = IngestionLoader.readRound(
        _fake_list_files({'r': [_f('r/mb.txt')]}),
        _fake_read_file({'r/mb.txt': mb_text}),
        'local', 'r', [], True, 0, 12,
    )
    assert r_clip['current']['chars'] == 30
    assert r_clip['current']['clipped'] is True
    assert r_clip['current']['text'] == '日本語' * 4
    assert len(r_clip['current']['text']) == 12


def test_read_round_surfaces_per_file_error():
    def boom_read(provider, fileId):
        if fileId == 'r/bad.txt':
            raise RuntimeError('read blew up')
        return {'is_text': True, 'data': 'ok'}

    r_err = IngestionLoader.readRound(
        _fake_list_files({'r': [_f('r/bad.txt'), _f('r/good.txt')]}),
        boom_read, 'local', 'r', [], True, 0,
    )
    assert r_err['current']['source'] == 'r/bad.txt'
    assert r_err['current']['error'] == 'read blew up'
    assert 'text' not in r_err['current']
